src/main/java/htsjdk/samtools/SamReaderFactory.java

```diff
@@ -311,7 +311,24 @@ public SamReader open(final SamInputResource resource) {
                     if (SamStreams.isBAMFile(bufferedStream)) {
                         if (sourceFile == null || !sourceFile.isFile()) {
                             // Handle case in which file is a named pipe, e.g. /dev/stdin or created by mkfifo
-                            primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                            if (indexFile!=null) {
+                                primitiveSamReader = new BAMFileReader(bufferedStream, indexFile, false, asynchronousIO, validationStringency, this.samRecordFactory);
+                            } else {
+                                final SeekableStream indexSeekable = indexMaybe == null ? null : indexMaybe.asUnbufferedSeekableStream();
+                                // do not close bufferedStream, it's the same stream we're getting here.
+                                SeekableStream sourceSeekable = data.asUnbufferedSeekableStream();
+                                if (null != sourceSeekable && null != indexSeekable) {
+                                    // let's make sure the user can seek in this file.
+                                    // need to return to the beginning because it's the same stream we used earlier
+                                    // and read a bit from.
+                                    sourceSeekable.seek(0);
```

Comment:
I think that's because the other places were written before this patch and so they expect the non-seekable stream to start from wherever isBAMFile leaves it. You can see in the [previous version of the code](https://github.com/samtools/htsjdk/blob/8e7fe7a0b204096c6d0736d473f20fb144908572/src/main/java/htsjdk/samtools/SamReaderFactory.java#L314) how the stream is passed to this variant of the constructor without rewinding. The seekable code path, instead, expects the stream to start from the beginning.